File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/D.Полиномиальный_хеш/solution.py
import sys


def solution(q: int, m: int, s: str) -> None:
    """
    Time Complexity O(s)
    Memory Complexity O(1)
    """
    h: int = 0
    for idx in range(len(s)):
        ch: str = s[idx]
        h += ord(ch) % m
        if idx != len(s) - 1:
            h = h * q % m
    sys.stdout.write(str(h % m) + '\n')


if __name__ == '__main__':
    a: int = int(sys.stdin.readline().rstrip())
    m: int = int(sys.stdin.readline().rstrip())
    s: str = sys.stdin.readline().rstrip()
    solution(a, m, s)

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/D.Полиномиальный_хеш/solution.cpp
#include <iostream>
#include <string>


void solution(int& q, int& m, std::string& s) {
    /*
    Time Complexity O(s)
    Memory Complexity O(1)
    */
    long long h = 0;
    for (int idx = 0; idx < s.size(); ++idx) {
        char ch = s[idx];
        h += (int)ch % m;
        if (idx != s.size() - 1) {
            h = (h % m * q % m) % m;
        }
    }
    std::cout << h % m << "\n";
}


int main() {
    int a, m;
    std::cin >> a >> m;
    std::string s;
    std::cin >> s;
    solution(a, m, s);
}